server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2773,4 +2773,64 @@ private static void trimUnsafeCommits(EngineConfig engineConfig) throws IOExcept
         final long minRetainedTranslogGen = Translog.readMinTranslogGeneration(translogPath, translogUUID);
         store.trimUnsafeCommits(globalCheckpoint, minRetainedTranslogGen, engineConfig.getIndexSettings().getIndexVersionCreated());
     }
+
+    /**
+     * Restores the live version map and local checkpoint of this engine using documents (including soft-deleted)
+     * after the local checkpoint in the safe commit. This step ensures the live version map and checkpoint tracker
+     * in sync with Lucene commit.
+     */
+    private void restoreVersionMapAndCheckpointTracker(DirectoryReader directoryReader) throws IOException {
+        final IndexSearcher searcher = new IndexSearcher(directoryReader);
+        searcher.setQueryCache(null);
+        final Query query = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getLocalCheckpoint() + 1, Long.MAX_VALUE);
+        final Weight weight = searcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
+        for (LeafReaderContext leaf : directoryReader.leaves()) {
+            final LeafReader reader = leaf.reader();
+            final Scorer scorer = weight.scorer(leaf);
+            if (scorer == null) {
+                continue;
+            }
+            final DocIdSetIterator docIdSetIterator = scorer.iterator();
+            final NumericDocValues seqNoDV = Objects.requireNonNull(reader.getNumericDocValues(SeqNoFieldMapper.NAME));
+            final NumericDocValues primaryTermDV = Objects.requireNonNull(reader.getNumericDocValues(SeqNoFieldMapper.PRIMARY_TERM_NAME));
+            final NumericDocValues tombstoneDV = reader.getNumericDocValues(SeqNoFieldMapper.TOMBSTONE_NAME);
+            final NumericDocValues versionDV = reader.getNumericDocValues(VersionFieldMapper.NAME);
+            int docId;
+            while ((docId = docIdSetIterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                if (primaryTermDV.advanceExact(docId) == false) {
+                    continue; // skip children docs which do not have primary term
+                }
+                final long primaryTerm = primaryTermDV.longValue();
+                if (seqNoDV.advanceExact(docId) == false) {
+                    throw new IllegalStateException("seq_no not found for doc_id=" + docId);
+                }
+                final long seqNo = seqNoDV.longValue();
+                localCheckpointTracker.markSeqNoAsCompleted(seqNo);
+                final boolean isTombstone = tombstoneDV != null && tombstoneDV.advanceExact(docId);
+                final FieldsVisitor fields = new FieldsVisitor(false, SourceFieldMapper.NAME);
```

Comment:
do we need to take RECOVERY_SOURCE into account as well (i.e. like in LuceneChangesSnapshot)?

Reply:
We only need the _id.